### ISSUE TITLE
INT8 block-quantized Qwen weights (llama.cpp-style Q8) — blocked on integer tensors in AutoDiff (docs/INTEGERS.md)

### ISSUE THREAD

**Issue by khurram-uworx** [khurram-uworx/Nivara#416]
## Summary

Captured from the QWEN-PERF **Stretch** row and the `docs/QWEN.md` performance
section: llama.cpp-style **INT8 block-quantized weights** (Q8) for the Qwen
decode path — ~4× weight-traffic cut, ~65+ tok/s memory-bandwidth ceiling at
Qwen2.5-0.5B shapes, at a fidelity cost.

**This is a blocked/dependency-gated item, not an open work item.**

## Blocker

Nivara's AutoDiff domain is constrained to floating-point types:

```csharp
where T : struct, IFloatingPointIeee754<T>
```

Integer tensors (I32/I64 — and by extension I8 quantization) are excluded at
compile time, so a quantized-weight inference path needs integer support
first. `docs/INTEGERS.md` documents the constraint, the SafeTensorsLoader
widening behavior, and the alternatives considered (including a
quantization-aware loader — **5c**, marked out of scope). Quantization
approaches can be addressed once integer support in AutoDiff is taken care of.

## Relationship to other items

- #390 — GGUF backend (a GGUF file may carry quantized weights; the format
  contract is model-loading-agnostic).
- #387 / #391 — BF16-in-memory SIMD-dot / revisit when `Vector<BFloat16>` SIMD
  lands: the narrower-dtype family this item is the 8-bit member of.

## Suggested direction (when unblocked)

- Block-scale dequantization (llama.cpp-style group quantization) widen-in-register
  per dot, mirroring the existing `WidenPrimitives`/bf16 widen path precedent.
- Quality gates: tolerance-bounded parity vs the F32 model (the existing
  QwenInstructParityTests logits envelope is the reference).

## Acceptance

- [ ] Unblocking prerequisite (integer tensors in AutoDiff) exists and is tracked.
- [ ] Q8 decode path measured vs F32 on the same machine (traffic, tok/s,
      fidelity delta).
- [ ] Parity fixture gate green.

Low priority / future; no prescriptive kernel work at this time.

**Comment by khurram-uworx** [comment]
# ChatGPT

If your goal is to run Qwen 2.5 0.5B-style GGUF quants efficiently, I would **not** generalize your library from

```csharp
where T : IFloatingPointIeee754<T>
```

to “all integer tensors.”

Instead, I would introduce a **quantized-weight representation and a small set of integer-oriented kernels**.

The important bit is that **Q4 isn't really an `int4` tensor in the .NET generic-numeric sense**.

### First: Ollama's Qwen 0.5B

If you're using the usual Ollama Qwen 2.5 0.5B model, check the exact model tag because Ollama can have different quantizations. Qwen 2.5 GGUF distributions commonly use `Q4_K_M`; llama.cpp's own tooling describes `Q4_K_M` as ~4.58 bits/weight, rather than literally 4 bits/weight. ([[GitHub](https://github.com/ggml-org/llama.cpp/blob/master/tools/quantize/quantize.cpp?utm_source=chatgpt.com)][1])

And `Q4_K` is structured roughly as:

```text
super-block
 ├── scales/mins
 └── packed 4-bit weights
```

with 8 blocks × 32 weights, and the effective storage is about 4.5 bits/weight. ([[GitHub](https://github.com/ggml-org/llama.cpp/wiki/Tensor-Encoding-Schemes?utm_source=chatgpt.com)][2])

So this:

```text
weight = int4
```

isn't quite the right abstraction.

It's more like:

```text
packed uint4 + per-block scale/minimum
```

and the important computation is generally something resembling:

```text
dot(x, dequant(q))
```

rather than:

```text
Tensor<int4>.Multiply(...)
```

---

# What I'd do in your .NET library

I'd keep your existing floating-point tensor abstraction:

```csharp
where T : IFloatingPointIeee754<T>
```

and add **quantized kernels separately**.

For example conceptually:

```csharp
interface IQuantizedKernel
{
    // ...
}

readonly struct Q4K
{
    // packed data + scales/mins
}
```

Then your hot path becomes something like:

```text
Q4_K weights
       │
       │ unpack / scale
       ▼
 SIMD integer operations
       │
       │ accumulate
       ▼
 float accumulation
```

rather than trying to make `TensorPrimitives` understand an `Int4` numeric type.

That distinction will save you a **lot** of pain.

---

## The SIMD type I'd target first: `int` / `uint`

For your implementation, **`byte`, `sbyte`, `short`, `ushort`, `int`, `uint` are the useful integer SIMD types**.

But for Q4 kernels I'd overwhelmingly expect you to use:

```csharp
Vector256<byte>
Vector256<sbyte>
Vector256<int>
Vector256<float>
```

and/or their `Vector512` equivalents.

You don't actually want a SIMD `Vector<T>` containing `int4`.

There is no useful native:

```csharp
Vector512<int4>
```

abstraction.

Instead, packed nibbles are unpacked into something like:

```text
packed:
  [abcd][efgh] ...

          ↓

unpacked:
  [a,b,c,d,e,f,g,h]
```

usually into byte/short/int lanes.

---

# And this is where TensorPrimitives becomes less useful

Your current design is probably excellent for F32:

```csharp
TensorPrimitives.Multiply(...)
TensorPrimitives.Add(...)
TensorPrimitives.Dot(...)
```

because the data already has a native numeric representation.

For Q4:

```text
packed Q4
    ↓
unpack
    ↓
scale / zero-point
    ↓
multiply
    ↓
accumulate
```

The **unpack/dequantization is itself part of the kernel**.

You don't want:

```csharp
TensorPrimitives.Dequantize(q4, floatBuffer);
TensorPrimitives.Dot(floatBuffer, activations);
```

because you've just thrown away a huge part of the benefit of quantization.

You want approximately:

```text
for each quantization block:
    decode q4
    SIMD multiply against activation
    accumulate
    apply scale
```

with the float conversion delayed as much as possible.

---

# The really interesting kernel is GEMV

For a 0.5B model on a laptop, your most important operation is probably effectively:

```text
y = W x
```

where:

```text
W = Q4
x = F32
y = F32
```

So your ideal kernel is:

```text
Q4 weights × F32 activations → F32 accumulation
```

not a generic:

```text
Tensor<T> × Tensor<T>
```

kernel.

Conceptually:

```text
             Q4 packed weights
                    │
             unpack 4-bit
                    │
              int8 / int16
                    │
                    ├──── multiply ──── activation F32
                    │
                    ▼
                F32 sum
                    │
                    ▼
                   y
```

There are more sophisticated approaches, too: integerizing the activations and doing integer dot products, depending on ISA and quantization scheme.

---

# Don't overlook `Q4_0`

If you're implementing your own inference runtime, I would actually consider supporting **Q4_0 before Q4_K_M**.

`Q4_0` is considerably easier to reason about:

```text
block of 32 weights

scale: float
weights: 32 × 4-bit
```

whereas K-quants introduce multiple scales/minimums and more complicated packing.

`Q4_K_M` is better as a production target, but it isn't necessarily the best **first quantized kernel**.

llama.cpp supports a whole family of quantization formats, including Q4_0, Q4_1, Q4_K, Q5_K, Q6_K, Q8_0, etc. ([[GitHub](https://github.com/ggml-org/llama.cpp/blob/master/src/llama-quant.cpp?utm_source=chatgpt.com)][3])

I'd potentially structure your implementation like:

```text
Phase 1
  F32                  ✓ already done
  Q4_0 GEMV            ← implement first

Phase 2
  Q8_0 GEMV
  Q4_K_M GEMV

Phase 3
  Q4_K_M optimized
  AVX2
  AVX-512 / Vector512
  ARM64 NEON

Phase 4
  other quantization formats
```

---

# One more important architectural point

I **wouldn't add `int` support to the entire TensorPrimitives-based API just because you need quantization.**

For example, this is probably unnecessary:

```csharp
Tensor<int>
Tensor<byte>
Tensor<short>
Tensor<float>
...
```

with every operation duplicated across them.

Instead:

```text
                  Tensor<T>
                     │
        ┌────────────┴────────────┐
        │                         │
 Floating point             Quantized tensors
        │                         │
 IFloatingPoint...        Q4_0 / Q4_K / Q8_0
        │                         │
 TensorPrimitives          specialized kernels
```

This is much closer to how an inference engine actually wants to think.

---

## And there's a particularly nice optimization available to you

You said you already use:

> `Vector512/Vector using recommended approaches`

That's good.

I'd make your quantized implementation **ISA-layered**, rather than tying the algorithm to `Vector512`.

Something like:

```text
Q4MatVec
   │
   ├── AVX-512 implementation
   │
   ├── AVX2 implementation
   │
   ├── AdvSimd implementation
   │
   └── scalar fallback
```

with the public kernel remaining something like:

```csharp
Q4MatVec(
    ReadOnlySpan<byte> weights,
    ReadOnlySpan<float> activations,
    Span<float> output);
```

The runtime chooses the implementation.

That's also broadly consistent with the .NET SIMD guidance: `TensorPrimitives` sits at a higher level, while `Vector128/256/512` and platform intrinsics provide progressively more specialized control. ([[GitHub](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/vectorization-guidelines.md?utm_source=chatgpt.com)][4])

---

# My recommendation for your project

Given what you've described, I would **not spend time implementing generic integer TensorPrimitives support yet**.

I'd do:

```text
Existing
────────
F32 tensors
IFloatingPointIeee754<T>
TensorPrimitives
Vector512 where appropriate

Add
────────
QuantizedTensor / QuantizedWeight abstraction

Q4_0
  ↓
Q4 GEMV F32 activations → F32 output
  ↓
SIMD unpack + multiply/accumulate

Then
────────
Q4_K_M
```

And importantly, **keep activations F32 initially**.

You can get a very substantial memory-bandwidth win from quantizing the weights without having to turn your whole inference pipeline into integer arithmetic.

The big mental model is:

> **You don't need “integer tensors.” You need “kernels capable of consuming packed quantized weights.”**

That's a much smaller — and IMO much more tractable — addition to your library.

If you show me **the current signature/implementation of one of your GEMM/GEMV kernels**, I can suggest exactly how I'd refactor it to support `Q4_0`/`Q4_K_M` while retaining your existing `TensorPrimitives` + `Vector512` architecture.

[1]: https://github.com/ggml-org/llama.cpp/blob/master/tools/quantize/quantize.cpp?utm_source=chatgpt.com "llama.cpp/tools/quantize/quantize.cpp at master · ggml-org/llama.cpp · GitHub"
[2]: https://github.com/ggml-org/llama.cpp/wiki/Tensor-Encoding-Schemes?utm_source=chatgpt.com "Tensor Encoding Schemes · ggml-org/llama.cpp Wiki · GitHub"
[3]: https://github.com/ggml-org/llama.cpp/blob/master/src/llama-quant.cpp?utm_source=chatgpt.com "llama.cpp/src/llama-quant.cpp at master · ggml-org/llama.cpp · GitHub"
[4]: https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/vectorization-guidelines.md?utm_source=chatgpt.com "runtime/docs/coding-guidelines/vectorization-guidelines.md at main · dotnet/runtime · GitHub"
